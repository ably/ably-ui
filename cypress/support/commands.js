Cypress.Commands.add(
  "compareComponents",
  ({
    reactComponentUrl,
    viewComponentUrl,
    beforeScreenshot = () => {},
    selector,
    componentName = Cypress.mocha.getRunner().suite.ctx.test.parent.title,
  }) => {
    let reactScreenshotPath = null;
    let vwScreenshotPath = null;

    cy.exec(`mkdir -p cypress/screenshots/component-parity/results`, {
      log: false,
    });

    const diffPath = `cypress/screenshots/component-parity/results/${componentName}.png`;

    // If you are wondering why this isn't Promise.all or async/await have a look at
    // https://docs.cypress.io/guides/core-concepts/introduction-to-cypress.html#Commands-Are-Not-Promises

    cy.visit(reactComponentUrl);

    beforeScreenshot();

    cy.get(selector)
      .screenshot(`${componentName}.react`, {
        onAfterScreenshot: (_, props) => (reactScreenshotPath = props.path),
      })
      .then(() => {
        cy.visit(viewComponentUrl);
        beforeScreenshot();

        cy.get(selector)
          .screenshot(`${componentName}.vw`, {
            onAfterScreenshot: (_, props) => (vwScreenshotPath = props.path),
          })
          .then(() => {
            cy.exec(
              `cypress/support/diff.js ` +
                `--pathA="${reactScreenshotPath}" ` +
                `--pathB="${vwScreenshotPath}" ` +
                `--target="${diffPath}" ` +
                `--threshold="0.005" `
            ).then((result) => {
              cy.exec(`rm "${reactScreenshotPath}" "${vwScreenshotPath}"`, {
                log: false,
              });

              assert.isTrue(result.stdout === "Yay", "Screenshots match");
            });
          });
      });
  }
);
