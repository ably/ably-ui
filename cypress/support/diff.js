#! /usr/bin/env node

// Used by the compareComponents command

const Blink = require("blink-diff");
const argv = require("yargs").argv;

const diff = new Blink({
  imageAPath: argv.pathA,
  imageBPath: argv.pathB,
  imageOutputPath: argv.target,
  threshold: 0.05,
});

diff.run((error, result) => {
  // We read stdout/stderr - this script is executed interactively by cypress
  if (error) {
    console.error(error);
  } else {
    console.log(diff.hasPassed(result.code) ? "Yay" : "Nay");
  }
});
