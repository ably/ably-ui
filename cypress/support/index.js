// ***********************************************************
// This example support/index.js is processed and
// loaded automatically before your test files.
//
// This is a great place to put global configuration and
// behavior that modifies Cypress.
//
// You can change the location of this file or turn off
// automatically serving support files with the
// 'supportFile' configuration option.
//
// You can read more here:
// https://on.cypress.io/configuration
// ***********************************************************

import "cypress-plugin-snapshots/commands";
import "./commands";

// Cypress has a bug that breaks snapshot tests when running the whole suite
// Follow https://github.com/cypress-io/cypress/issues/3090
// Recommended workaround below:
export const fixSnapshotSpec = (filename) => () => {
  const path = require("path");
  // const relative = filename.substr(1); // removes leading "/"
  const projectRoot = Cypress.config("projectRoot");
  const absolute = path.join(projectRoot, filename);

  Cypress.spec = {
    absolute,
    name: path.basename(filename),
    filename,
  };
};
