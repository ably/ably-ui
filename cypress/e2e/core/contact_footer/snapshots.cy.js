import { fixSnapshotSpec } from "../../../support";

beforeEach(fixSnapshotSpec(__filename));

describe("Chat button state", () => {
  beforeEach(() => {
    cy.visit("/components/contact-footer");
  });

  it("the button is inactive without Hubspot widget", () => {
    cy.get("[data-id='open-chat-widget']").toMatchSnapshot();
  });
});
