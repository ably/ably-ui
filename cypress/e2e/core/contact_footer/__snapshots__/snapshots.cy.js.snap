exports[`Chat button state > the button is inactive without Hubspot widget #0`] = `
<button
  type="button"
  class="ui-btn-secondary self-start mt-16"
  disabled=""
  data-id="open-chat-widget"
  data-enabled-label="Start a live chat"
  data-disabled-label="Live chat unavailable"
>
  Live chat unavailable
</button>
`;
